ipa-getkeytab enhancements

martbab opened the pull request:
This PR implements '-H' and '-Y' options mentioned in
https://fedorahosted.org/freeipa/ticket/6409 along with the ability to specify
CA cert on the command line (which proved useful during the work on installer
refactoring).

Since my C skills are not at the level I would like them to be it would be nice
if you point out even the tiniest mistakes, risky code or non-idiomatic usage.

Also the test case `test_retrieval_using_plain_ldap` fails due to unsuccesful
simple bind. I wanted to implement StartTLS for simple binds over ldap://, but
I get the following errors in dirsrv error log:

        [01/Nov/2016:10:44:52.395126000 +0000] connection - conn=883 fd=135
        Incoming BER Element was 3 bytes, max allowable is 209715200 bytes.
        Change the nsslapd-maxbersize attribute in cn=config to increase.

I guess there is something fishy with the way I initialize the StartTLS
session. I would appreciate your help with it.

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -712,6 +717,8 @@ int main(int argc, const char *argv[])
 	    { "bindpw", 'w', POPT_ARG_STRING, &bindpw, 0,
               _("LDAP password"), _("password to use if not using kerberos") },
+	    { "cacert", 'c', POPT_ARG_STRING, &ca_cert_file, 0,
+            _("Path to the IPA CA certificate"), _("IPA CA certificate")},
```
indentation does not follow the style of the rest of the code here

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -137,26 +158,48 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
+    int port = 0;
+    char *scheme = NULL;
+
+    if (!mech) {
```
I do not understand why you are using the mech to determine the URI scheme, these are unrelated things.

martbab replied:
That is probably some leftover from previous incarnations of the patch.  If I manage to make StartTLS work this may actually be irrelevant since we may use ldap://$FQDN and have secure connection.

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -137,26 +158,48 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
+    else {
+        return 1;
+    }
+    url_len = snprintf(url,0,"%s://%s:%d",scheme,servername,port) +1;
```
please just use asprintf() in here.

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -804,6 +890,14 @@ int main(int argc, const char *argv[])
+        ret = ipa_server_to_uri(server, sasl_mech, &ldap_uri);
+        if (ret) {
+            fprintf(stderr, _("Incompatible combination of server name and "
+                              "SASL bind mechanism\n"));
```
This error message looks wrong.
ipa_server_to_uri() currently fails only if the mechanism is not recognized (but that one is checked earlier already), or if string allocation fails.

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -837,7 +931,7 @@ int main(int argc, const char *argv[])
 	}
 
-	if (NULL == bindpw) {
+	if (NULL == bindpw && strncmp(sasl_mech, LDAP_SASL_GSSAPI, 79) == 0) {
```
79 ?

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -196,6 +196,21 @@ static int ipa_ldap_init(LDAP **ld, const char *ldap_uri)
+    int rc = LDAP_SUCCESS;
+    char *startswith_ldap = NULL;
+
+    startswith_ldap = strstr("ldap://", ldap_uri);
```
ldap:// can be only at the start of the string, pleae use strncmp(ldap_uri, "ldap://", 7)

martbab commented:
Thank you for review @simo5 . I have fixed the issues and reworked the LDAP initialization and binding logic a bit to clean it up. It produced green tests for me. I have also updated the command man page as I missed that during initial work.

martbab commented:
Bump for review, this is prerequisite to further installer refactoring work.

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -135,61 +158,86 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
-	}
-	sprintf(url,"%s://%s:%d",scheme,servername,port);
-	int rc = ldap_initialize(ld, url);
+    url_len = asprintf(&url,"%s://%s:%d",scheme,servername,port);
```
space after comma

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -135,61 +158,86 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
 
-	free(url);
-	return rc;
+    if (url_len == -1){
```
space after close parens

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -135,61 +158,86 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
+    int tls_hard = LDAP_OPT_X_TLS_HARD;
+    int tls_demand = LDAP_OPT_X_TLS_DEMAND;
+    char *ldap = "ldap://";
+    char *ldaps = "ldaps://";
```
there is no need to use an actual variable, beside these would have to be defined as const char as the strings are constant.

Use defines:
```#define SCHEMA_LDAP "ldap://"```
```#define SCHEMA_LDAPS "ldaps://"```
Then for lenght you can use sizeof(SCHEMA_LDAP) -1

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -135,61 +158,86 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
-            fprintf(stderr, _("Unable to init for ldap(389) connection\n"));
             return ret;
         }
+
```
don't add empty line

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -221,21 +274,26 @@ static int ipa_ldap_bind(const char *server_name, krb5_principal bind_princ,
                                &bv, NULL, NULL, NULL);
         if (ret != LDAP_SUCCESS) {
             fprintf(stderr, _("Simple bind failed\n"));
-            goto done;
```
This looks wrong, if bind fails we should get out

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -221,21 +274,26 @@ static int ipa_ldap_bind(const char *server_name, krb5_principal bind_princ,
                             ldap_err2string(ret), ret, msg ? msg : "");
             goto done;
+            }
         }
```
please make sure the close parens matches the same indentation of the open one at all times

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -221,21 +274,26 @@ static int ipa_ldap_bind(const char *server_name, krb5_principal bind_princ,
-                                           ldap_sasl_interact, bind_princ);
+        if (strcmp(mech, LDAP_SASL_EXTERNAL) == 0) {
+            ret = ldap_sasl_bind_s(ld, NULL, LDAP_SASL_EXTERNAL,
+                               NULL, NULL, NULL, NULL);
```
bad indent, sshould start under 'ld'

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -783,7 +857,34 @@ int main(int argc, const char *argv[])
+        fprintf(stderr, _("Cannot specify both SASL mechanism "
+                          "and bind DN simultaneously.\n"));
+		if (!quiet)
+			poptPrintUsage(pc, stderr, 0);
```
incorrect indentation

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -783,7 +857,34 @@ int main(int argc, const char *argv[])
+        fprintf(stderr, _("Cannot specify server and LDAP uri "
+                          "simultaneously.\n"));
+		if (!quiet)
+			poptPrintUsage(pc, stderr, 0);
```
incorrect indentation

simo5 commented on client/man/ipa-getkeytab.1:
```diff
@@ -96,7 +96,17 @@ Use this password for the key instead of one randomly generated.
+The LDAP password to use when not binding with Kerberos. \fB\-D\fR and \fB\-w\fR can not be used together with \fB\-Y\fR.
+.TP
+\fB\-\-cacert\fR
+The path to IPA CA certificate used to validate LDAPS/STARTTLS connection.
```
-> path to 'the' IPA CA certificate
-> connections (plural)

simo5 commented on client/man/ipa-getkeytab.1:
```diff
@@ -96,7 +96,17 @@ Use this password for the key instead of one randomly generated.
+Defaults to /etc/ipa/ca.crt
+.TP
+\fB\-H, \-\-ldapuri\fR
+LDAP URI. If ldap:// is specified, STARTTLS session is initiated by default. Can not be used with \fB\-s\fR.
```
remove 'session'

simo5 commented on client/man/ipa-getkeytab.1:
```diff
@@ -96,7 +96,17 @@ Use this password for the key instead of one randomly generated.
+LDAP URI. If ldap:// is specified, STARTTLS session is initiated by default. Can not be used with \fB\-s\fR.
+.TP
+\fB\-Y, \-\-mech\fR
+SASL mechanism to use if fB\-D\fR and \fB\-w\fR are not used. Choose between GSSAPI and EXTERNAL.
```
-> s/Choose between GSSAPI and/Use either GSSAPI or/

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -135,61 +158,86 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
-        ssl = LDAP_OPT_X_TLS_HARD;;
-        ret = ldap_set_option(ld, LDAP_OPT_X_TLS, &ssl);
+    }
+    else if (strncmp(ldap_uri, ldaps, strlen(ldaps)) == 0) {
```
} else if,
don't go to a new line wth the else

martbab commented:
Another bump for review.

simo5 commented on client/ipa-getkeytab.c:
```diff
@@ -135,62 +161,82 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
-	}
-	sprintf(url,"%s://%s:%d",scheme,servername,port);
-	int rc = ldap_initialize(ld, url);
+    url_len = asprintf(&url,"%s%s:%d", SCHEMA_LDAP, servername, port);
```
space after comma

martbab commented:
Done.

martbab commented:
Fixed upstream
master:
https://fedorahosted.org/freeipa/changeset/294fc3dc5645eeb7942908c3e351c06aa0af329e
https://fedorahosted.org/freeipa/changeset/0c68c27e51c2a30265a760382d7d4fab7d21937b
https://fedorahosted.org/freeipa/changeset/8480d0e3333f6813439e7b3321a0e33ce80d30f1
https://fedorahosted.org/freeipa/changeset/2725e440bf1e4930f9b1d19223424bcb0d4b7066
